### ISSUE TITLE
"Warning - couldn't find a Python 3.8 installation."

### ISSUE THREAD

**Issue by dirkk0** [christofmuc/KnobKraft-orm#84]
First of all: thanks for writing an OSing this! Much appreciated!

I run this on Mac OSX Cataline, I have Python 3.8 installed, still I get this message:
![image](https://user-images.githubusercontent.com/625394/106740977-66d02800-661b-11eb-94a3-0716bf714a81.png)
Any idea how to fix this?

Thanks,
  Dirk

**Comment by christofmuc** [comment]
Hi Dirk, let's find out where the installation of Python is. The Orm looks in 

     /Library/Frameworks/Python.framework/Versions/3.8

and then in 

    /usr/local/opt/python3/Frameworks/Python.framework/Versions/3.8

which I think are the two valid installation directories, at least if homebrew and Xcode agree.

Can you run a "which python" to see? Maybe a symlink will suffice to get it running. Do you remember how you installed Python 3.8?

**Comment by dirkk0** [comment]
I think I did ```brew reinstall python@3.8``` since ```brew install python``` installed Python 3.9.

My python is in ```/usr/bin/python3```.

**Comment by christofmuc** [comment]
Thank you! I have no Mac here right now to check, but I think what is important is not only the Python binary, but also where the libraries are. Can you check the two paths above and see if anything is in either of the two places?

**Comment by dirkk0** [comment]
The above paths don't work since there is only the default 2.7 Framework there.
[This](https://www.xspdf.com/resolution/54674256.html) helped me find me the path. For my machine it seems to be:
```/usr/local/Cellar/python@3.8/3.8.7/Frameworks/Python.framework```

Does that make sense?

**Comment by christofmuc** [comment]
This page documents the behavior of homebrew quite nicely, even it seems overly complex to me:

https://docs.brew.sh/Homebrew-and-Python

So it seems we are looking for the value of 

(brew --prefix)/lib/pythonX.Y

The directory should be there and should have a subdirectory `site-packages`, as far as I can see. The page above says this should be /usr/local/lib/python3.y/site-packages. This does not match my experience, though, and sadly I don't have a Mac right now to test this. Maybe the "reinstall" command did omit something that would have normally linked the versioned Frameworks directory correctly.

It seems Cellar is the package directory for all things installed by brew, but it doesn't seem this is the directory from where to run it from. Maybe there are other install mechnanisms that will put Python 3.8 into one of the two places assumed by the Orm? Like this installer https://www.python.org/ftp/python/3.8.7/python-3.8.7-macosx10.9.pkg

Another "quick" solution for you even if daring would be to compile yourself and adapt the top level CMakeLists.txt to just use the Python 3.9. you already have.. can't think right now why that wouldn't work.

Mid-termin solution is that I make the path configurable via an environment variable so you can override it should the default package location not work.

**Comment by dirkk0** [comment]
Thanks again for your support!

I successfully (?) compiled the application (meaning it came to an end with all green messages):
```[100%] Built target KnobKraftOrm```

 but it couldn't start with this error:

    Check with the developer to make sure KnobKraftOrm works with this version of macOS.
    You may need to reinstall the application. Be sure to install any available updates for the application and macOS.

In the report it says: 

    Dyld Error Message:
      Library not loaded: @rpath/Python3.framework/Versions/3.8/Python3
      Referenced from: /Users/USER/*/KnobKraftOrm.app/Contents/MacOS/KnobKraftOrm
      Reason: image not found

( [full report](https://pastebin.com/cv1LLzPr) )

I have 3.8.2 installed, btw.

Where would I edit the Python in the CMake?

**Comment by christofmuc** [comment]
Phew. Python is included via the library Pybind11, https://github.com/pybind/pybind11, which normally does a good job. The include is in the top level CMakeLists.txt (add_subdirectory(third_party/pybind) or similar). 

Here is an old issue with this error message, not sure if it applies: https://github.com/Homebrew/legacy-homebrew/issues/29590

I don't use virtualenv in the software, so normally when it links it should also find its libraries? Not sure what the @rpath reference is in the error message. On Linux I know the LDPATH which needs to be set. You could also have a look at lines 210ff in GenericAdaptation.cpp, which is where the path for Python is set in code.

**Comment by christofmuc** [comment]
Here is a bit more info on the rpath thing: https://stackoverflow.com/questions/29809069/dyld-library-not-loaded-on-os-x

You could try to build the Mac version the same way I do it in the azure-pipelines.yml file:

    cmake --build builds/release --target package -j8 

The --target package will somehow assemble everything needed to run in a DMG file?

**Comment by dirkk0** [comment]
Yes, this did create the DMG, but no, still the same error.
I'll try to go the non-homebrew route now, and will install the 'official' Python3.

**Comment by dirkk0** [comment]
Somehow things got better and worse at the same time. :)
The good news: the original client I downloaded here started to work (I guess because I installed the 'correct' python).
So that's very cool!

The bad (not really) news is that the compilation created a broken binary (like, the app had a strike-through).
EDIT: after I ran the original app, the broken compilation showed up unbroken ... I think?
This sounds weird so I'll rather recompile from a fresh start to make sure everything is working and I didn't screw up.
I'll post the results here.
